include/nbl/builtin/hlsl/math/functions.hlsl

```diff
@@ -296,50 +313,66 @@ bool partitionRandVariable(float leftProb, NBL_REF_ARG(float) xi, NBL_REF_ARG(fl
     // This is all 100% correct taking into account the above NEXT_ULP_AFTER_UNITY
     xi -= pickRight ? leftProb : 0.0;
 
-    rcpChoiceProb = 1.0 / (pickRight ? (1.0 - leftProb) : leftProb);
+    rcpChoiceProb = NBL_FP64_LITERAL(1.0) / (pickRight ? (NBL_FP64_LITERAL(1.0) - leftProb) : leftProb);
     xi *= rcpChoiceProb;
 
     return pickRight;
 }
 
+namespace impl
+{
+template <typename T NBL_STRUCT_CONSTRAINABLE>
+struct conditionalAbsOrMax_helper;
 
-// TODO: make it work in C++, ignoring problem for now
-#ifdef __HLSL_VERSION
-// @ return abs(x) if cond==true, max(x,0.0) otherwise
-template <typename T NBL_FUNC_REQUIRES(is_scalar_v<T> || is_vector_v<T>)
-T conditionalAbsOrMax(bool cond, T x, T limit);
+// TODO: conditionalAbsOrMax_helper partial template specialization for signed integers
 
-template <>
-float conditionalAbsOrMax<float>(bool cond, float x, float limit)
+template <typename T> NBL_PARTIAL_REQ_TOP(concepts::FloatingPointLikeScalar<T>)
+struct conditionalAbsOrMax_helper<T NBL_PARTIAL_REQ_BOT(concepts::FloatingPointLikeScalar<T>) >
 {
-    const float condAbs = asfloat(asuint(x) & uint(cond ? 0x7fFFffFFu : 0xffFFffFFu));
-    return max(condAbs,limit);
-}
+    static T __call(bool cond, NBL_CONST_REF_ARG(T) x, NBL_CONST_REF_ARG(T) limit)
+    {
+        using UintOfTSize = unsigned_integer_of_size_t<sizeof(T)>;
+        const T condAbs = bit_cast<T>(bit_cast<UintOfTSize>(x) & (cond ? 0x7fFFffFFu : 0xffFFffFFu));
```

Comment:
the bitmasks only work for `float32_t`, construct them from `numeric_limits` or the `ieee754`  stuff